### PR TITLE
LwM2M Client LTE network registration status notify and automatic LwM2M engine pause/resume

### PR DESCRIPTION
**Pull request by jheiskan81**
Extended neighbor cell notification module to support also LTE network registration status notification.
LwM2M client sample enable LTE network status monitoring and automatically pause and resume LwM2M engine based on client connected status and LTE network status.

### CODE REVIEW

**Comment by jheiskan81** [comment]
@rlubos @SeppoTakalo @jarlamsa @VeijoPesonen Please review.

**Comment by jheiskan81** [comment]
@nordicjm Do not merge untill review is ready. If Upmerge is ready before that I do normal PR to nrf-sdk.

File: samples/nrf9160/lwm2m_client/src/main.c [hunk 1]
```diff
@@ -382,6 +392,29 @@ static void modem_connect(void)
...
+{
+	if (connected) {
+		LOG_INF("LTE connected");
+		if (lwm2m_stack_suspended) {
+			if (lwm2m_engine_resume() == 0) {
+				lwm2m_stack_suspended = false;
+			}
+		}
+	} else {
+		LOG_INF("LTE disconne disconnected");
```
Comment (VeijoPesonen) [inline samples/nrf9160/lwm2m_client/src/main.c]:
```suggestion
		LOG_INF("LTE disconnected");
```

File: subsys/net/lib/lwm2m_client_utils/Kconfig [hunk 2]
```diff
@@ -132,6 +132,9 @@ config LWM2M_CLIENT_UTILS_LOCATION_ASSIST_AGPS_BUF_SIZE
 	depends on LWM2M_CLIENT_UTILS_LOCATION_ASSIST_AGPS
 	default 4096
 
+config LWM2M_CLIENT_UTILS_NWK_REG_NOTIFICATION
+	bool "Enable LE Network registration status Notification"
```
Comment (VeijoPesonen) [inline subsys/net/lib/lwm2m_client_utils/Kconfig]:
```suggestion
	bool "Enable LTE Network registration status Notification"
```

File: include/net/lwm2m_client_utils.h [hunk 3]
```diff
@@ -172,6 +172,26 @@ void location_assist_cell_request_set(void);
...
+/**
+ * @brief LTE network registration update state change event callback.
+ *
+ * @param[in] connected Network registration status True or False
+ *
+ */
+typedef void (*lwm2m_lte_nwk_reg_update_state_cb_t)(bool connected);
+
+/**
+ * @brief LTE network registration update event order.
```
Comment (jarlamsa) [inline include/net/lwm2m_client_utils.h]: Suggestion: Register a callback for receiving LTE network status notifications.

File: include/net/lwm2m_client_utils.h [hunk 4]
```diff
@@ -172,6 +172,26 @@ void location_assist_cell_request_set(void);
...
+ *
+ * @param[in] cb A callback function to receive lte network state changes
```
Comment (jarlamsa) [inline include/net/lwm2m_client_utils.h]: Suggestion: cb A callback function to receive LTE network status notifications.

File: samples/nrf9160/lwm2m_client/src/main.c [hunk 5]
```diff
@@ -335,19 +344,20 @@ static void rd_client_event(struct lwm2m_ctx *client, enum lwm2m_rd_client_event
 
 	case LWM2M_RD_CLIENT_EVENT_DISCONNECT:
 		LOG_DBG("Disconnected");
+		lwm2m_client_connected = false;
 		break;
 
 	case LWM2M_RD_CLIENT_EVENT_QUEUE_MODE_RX_OFF:
 		LOG_DBG("Queue mode RX window closed");
 		break;
-
```
Comment (jarlamsa) [inline samples/nrf9160/lwm2m_client/src/main.c:343]: Nitpick: Don't remove empty line here.

File: samples/nrf9160/lwm2m_client/src/main.c (hunk 1 above)
Comment (jarlamsa) [inline samples/nrf9160/lwm2m_client/src/main.c]: Nitpick: LTE disconnected

File: samples/nrf9160/lwm2m_client/src/main.c [hunk 6]
```diff
@@ -382,6 +392,29 @@ static void modem_connect(void)
 	} while (ret < 0);
 }
 
+#if defined(CONFIG_LWM2M_CLIENT_UTILS_SIGNAL_MEAS_INFO_OBJ_SUPPORT)
```
Comment (VeijoPesonen) [inline samples/nrf9160/lwm2m_client/src/main.c]: This seems to be wrong.

File: samples/nrf9160/lwm2m_client/src/main.c [hunk 7]
```diff
@@ -382,6 +392,29 @@ static void modem_connect(void)
...
+		LOG_INF("LTE connected");
+		if (lwm2m_stack_suspended) {
+			if (lwm2m_engine_resume() == 0) {
+				lwm2m_stack_suspended = false;
+			}
+		}
+	} else {
+		LOG_INF("LTE disconne disconnected");
+		if (!lwm2m_stack_suspended && lwm2m_client_connected) {
+			/* Are we connected already */
```
Comment (jarlamsa) [inline samples/nrf9160/lwm2m_client/src/main.c]: This is a bit confusing comment, consider removing.

File: samples/nrf9160/lwm2m_client/src/main.c [hunk 8]
```diff
@@ -429,6 +462,10 @@ void main(void)
 		return;
 	}
 
+#if defined(CONFIG_LWM2M_CLIENT_UTILS_SIGNAL_MEAS_INFO_OBJ_SUPPORT)
```
Comment (VeijoPesonen) [inline samples/nrf9160/lwm2m_client/src/main.c]: This seems to be wrong too.

**Review by VeijoPesonen — CHANGES_REQUESTED** [review]

File: samples/nrf9160/lwm2m_client/src/main.c [hunk 9]
```diff
@@ -60,6 +60,11 @@ static int reconnection_counter;
 static struct k_sem lwm2m_restart;
 /* Enable session lifetime check for initial boot */
 static bool update_session_lifetime = true;
+static bool lwm2m_client_connected;
+
+#if defined(CONFIG_LWM2M_CLIENT_UTILS_SIGNAL_MEAS_INFO_OBJ_SUPPORT)
```
Comment (jarlamsa) [inline samples/nrf9160/lwm2m_client/src/main.c]: Why the suspend is now depending of the signal measurement info object that is used just for neighbor cell measurements?

File: subsys/net/lib/lwm2m_client_utils/Kconfig (hunk 2 above)
Comment (jarlamsa) [inline subsys/net/lib/lwm2m_client_utils/Kconfig]: Suggestion: "Enable LTE network registration status notifications"

**Comment by jheiskan81** [comment]
@VeijoPesonen Thanks for good findings I have been fixed those.

File: subsys/net/lib/lwm2m_client_utils/CMakeLists.txt [hunk 10]
```diff
@@ -14,8 +14,8 @@ zephyr_library_sources_ifdef(CONFIG_LWM2M_CLIENT_UTILS_DEVICE_OBJ_SUPPORT lwm2m/
 zephyr_library_sources_ifdef(CONFIG_LWM2M_CLIENT_UTILS_SECURITY_OBJ_SUPPORT lwm2m/lwm2m_security.c)
 zephyr_library_sources_ifdef(CONFIG_LWM2M_CLIENT_UTILS_SIGNAL_MEAS_INFO_OBJ_SUPPORT
 			     lwm2m/ecid_signal_meas_info.c)
-zephyr_library_sources_ifdef(CONFIG_LWM2M_CLIENT_UTILS_NEIGHBOUR_CELL_LISTENER
-			     lwm2m/neighbour_cell_info.c)
+zephyr_library_sources_ifdef(CONFIG_LWM2M_CLIENT_UTILS
+			     lwm2m/lwm2m_lte_notification.c)
```
Comment (jarlamsa) [inline subsys/net/lib/lwm2m_client_utils/CMakeLists.txt:18]: Changing this dependency causes some issues as the `CONFIG_LWM2M_CLIENT_UTILS_NEIGHBOUR_CELL_LISTENER` is dependent of `CONFIG_LWM2M_CLIENT_UTILS_SIGNAL_MEAS_INFO_OBJ_SUPPORT`
The code shouldn't try to populate the neighbour cell information to object if the object itself isn't supported.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_lte_notification.c [hunk 11]
```diff
@@ -94,9 +98,57 @@ int lwm2m_update_signal_meas_objects(const struct lte_lc_cells_info *const cells
...
+{
+if (!event_handler_requested) {
+		event_handler_requested = true;
+		lte_lc_register_handler(lte_notification_handler);
+	}
+	lte_nwk_reg_cb = cb;
+	return 0;
+}
+
+static void lwm2m_lte_reg_handler_notifify(enum lte_lc_nw_reg_status nw_reg_status)
```
Comment (jarlamsa) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_lte_notification.c]: nitpick: notify

**Review by VeijoPesonen — APPROVED** [review]

**Review by jarlamsa — CHANGES_REQUESTED** [review]
The kconfig dependencies now cause issues. The network status notifications shouldn't be dependent of the signal measurement info objects nor should the measurement info objects be used when they aren't configured to be used.

File: samples/nrf9160/lwm2m_client/src/main.c (hunk 9 above)
Comment (jheiskan81) [reply samples/nrf9160/lwm2m_client/src/main.c]: I fix this already

File: subsys/net/lib/lwm2m_client_utils/Kconfig (hunk 2 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m_client_utils/Kconfig]: fixed.

File: subsys/net/lib/lwm2m_client_utils/CMakeLists.txt (hunk 10 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m_client_utils/CMakeLists.txt:18]: good point I change lwm2m_lte_notification() to use CONFIG_LWM2M_CLIENT_UTILS_NEIGHBOUR_CELL_LISTENER

**Comment by jheiskan81** [comment]
@jarlamsa Thanks for valuable feedback. I have fixed comment's and findings.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_lte_notification.c [hunk 12]
```diff
@@ -17,6 +17,10 @@
 
 #include <zephyr/logging/log.h>
 LOG_MODULE_REGISTER(lwm2m_neighbour_cell, CONFIG_LWM2M_CLIENT_UTILS_LOG_LEVEL);
```
Comment (jarlamsa) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_lte_notification.c]: Maybe change the module name as the module isn't just about ncell anymore.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_lte_notification.c [hunk 13]
```diff
@@ -17,6 +17,10 @@
...
+static void lte_notify_handler(const struct lte_lc_evt *const evt);
+
+static bool event_handler_requested;
```
Comment (jarlamsa) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_lte_notification.c]: These 2 should be in the `#ifdef(CONFIG_LWM2M_CLIENT_UTILS_NWK_REG_NOTIFICATION)`

**Review by jarlamsa — CHANGES_REQUESTED** [review]
2 small things only, otherwise looking good already.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_lte_notification.c (hunk 13 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_lte_notification.c]: No these 2 are needed for both ncell and Network registration notify.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_lte_notification.c (hunk 12 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_lte_notification.c]: true I can do it.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_lte_notification.c [hunk 14]
```diff
@@ -94,9 +98,57 @@ int lwm2m_update_signal_meas_objects(const struct lte_lc_cells_info *const cells
...
+#endif
+
+#if defined(CONFIG_LWM2M_CLIENT_UTILS_NWK_REG_NOTIFICATION)
+static lwm2m_lte_nwk_reg_update_state_cb_t lte_nwk_reg_cb;
+
+int lwm2m_lte_reg_handler_register(lwm2m_lte_nwk_reg_update_state_cb_t cb)
+{
+if (!event_handler_requested) {
+		event_handler_requested = true;
+		lte_lc_register_handler(lte_notify_handler);
```
Comment (jarlamsa) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_lte_notification.c]: If you would create a common init instead, you wouldn't need register protection boolean. Have another function to set the callback.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_lte_notification.c (hunk 13 above)
Comment (jarlamsa) [reply subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_lte_notification.c]: Oh yeah, true.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_lte_notification.c (hunk 14 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_lte_notification.c]: OK I create a common init.

**Comment by jheiskan81** [comment]
@jarlamsa I fx you comment- I realize that I can register event handler multiple time it is not problem. link control code detect if handler is already register it not allocate new. I remove boolean because there was not any advantage.

**Comment by jheiskan81** [comment]
I close PR make PR to main branch.

**Pull request closed without merging**